File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums = sorted(nums)
        N = len(nums)
        for i,a in enumerate(nums):
            if i > 0 and a == nums[i-1]:
                continue
            right = N - 1
            left = i + 1
            while(left < right):
                if a + nums[left] + nums[right] > 0:
                    right-=1
                elif a + nums[left] + nums[right] < 0:
                    left+=1
                else:
                    res.append([a,nums[left],nums[right]])
                    left+=1
                    while(nums[left] == nums[left - 1] and left < right):
                        left+=1
        return res

File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        int i = 0;
        Arrays.sort(nums);
        List<Integer> curr;
        while(i < nums.length - 2){
            if(nums[i] > 0){
                break;
            }
            int j = i + 1;
            int k = nums.length-1;
            while(j < k){
                if(nums[j] + nums[k] == -1* nums[i]){
                    result.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    while(j < k && nums[j+1] == nums[j]){
                        j+=1;
                    }
                    j+=1;
                    while(k > j && nums[k-1] == nums[k]){
                        k-=1;
                    }
                    k-=1;
                }
                else if(nums[j] + nums[k] > -1* nums[i]){
                    while(k > j && nums[k-1] == nums[k]){
                        k-=1;
                    }
                    k-=1;
                }
                else{
                    while(j < k && nums[j+1] == nums[j]){
                        j+=1;
                    }
                    j+=1;
                }
            }
            while(i < nums.length - 2 && nums[i+1] == nums[i]){
                i+=1;
            }
            i+=1;
        }
        return result;
    }
}